druid-shell/src/platform/x11/window.rs

```diff
@@ -314,64 +292,37 @@ impl Window {
-        match self.cairo_context.try_borrow() {
-            Ok(ctx) => match ctx.get_target().try_into() {
-                Ok(surface) => Ok(surface),
-                Err(err) => Err(Error::Generic(format!(
-                    "Window::cairo_surface ctx.try_into: {}",
-                    err
-                ))),
-            },
-            Err(err) => Err(Error::BorrowError(format!(
-                "Window::cairo_surface cairo_context: {}",
-                err
-            ))),
-        }
+    fn cairo_surface(&self) -> Result<XCBSurface, anyhow::Error> {
+        self.cairo_context
+            .try_borrow()
+            .context("Window::cairo_surface cairo_context")?
+            .get_target()
+            .try_into()
+            .map_err(|_| anyhow!("Window::cairo_surface try_into"))
     }
 
     fn size(&self) -> Result<Size, Error> {
-        match self.state.try_borrow() {
-            Ok(state) => Ok(state.size),
-            Err(err) => Err(Error::BorrowError(format!("Window::size state: {}", err))),
-        }
+        Ok(self.state.try_borrow().context("Window::size")?.size)
     }
 
     fn set_size(&self, size: Size) -> Result<(), Error> {
         // TODO(x11/dpi_scaling): detect DPI and scale size
-        let new_size = match self.state.try_borrow_mut() {
-            Ok(mut state) => {
-                if state.size != size {
-                    state.size = size;
-                    Some(size)
-                } else {
-                    None
-                }
-            }
-            Err(err) => {
-                return Err(Error::BorrowError(format!(
-                    "Window::set_size state: {}",
-                    err
-                )))
-            }
-        };
-        if let Some(size) = new_size {
-            let cairo_surface = self.cairo_surface()?;
-            if let Err(err) = cairo_surface.set_size(size.width as i32, size.height as i32) {
-                return Err(Error::Generic(format!(
-                    "Failed to update cairo surface size to {:?}: {}",
-                    size, err
-                )));
-            }
-            match self.handler.try_borrow_mut() {
-                Ok(mut handler) => handler.size(size),
-                Err(err) => {
-                    return Err(Error::BorrowError(format!(
-                        "Window::set_size handler: {}",
-                        err
-                    )))
-                }
-            }
+        let mut state = self.state.try_borrow_mut().context("Window::set_size")?;
+        if state.size != size {
+            state.size = size;
+            self.cairo_surface()?
+                .set_size(size.width as i32, size.height as i32)
+                .map_err(|status| {
+                    anyhow!(
+                        "Failed to update cairo surface size to {:?}: {}",
+                        size,
+                        status
+                    )
+                })?;
+            self.handler
```

Comment:
This changes the lifetime of the `state` borrow. In the previous implementation it gets dropped as quickly as possible and certainly before calling the `size` handler which might call a method on `WindowHandle` which will want to borrow `state`. At the very least we should use `mem::drop` or something to ensure we don't hold it when calling this handler.

Reply:
Good point, I've gone back to the old control flow.